### PR TITLE
Limit max number of update for workflow execution

### PR DESCRIPTION
**Pull request by alexshtin**
**What changed?**
Limit max number of update for workflow execution. Default is `2000`.

**Why?**
To prevent uncontrollable mutable state size grow.

**How did you test it?**
Existing tests.

**Potential risks**
No risks.

**Is hotfix candidate?**
No.

### CODE REVIEW

File: service/history/workflow/update/registry.go [hunk 1]
```diff
@@ -224,14 +233,21 @@ func (r *RegistryImpl) remover(id string) updateOpt {
...
+	if len(r.updates) >= r.maxInFlight() {
+		return serviceerror.NewResourceExhausted(
+			enumspb.RESOURCE_EXHAUSTED_CAUSE_CONCURRENT_LIMIT,
+			fmt.Sprintf("update concurrent in-flight limit has been reached (%v)", r.maxInFlight()),
+		)
+	}
+
+	if r.store.GetUpdatesCount(ctx) >= r.maxTotal() {
 		return serviceerror.NewResourceExhausted(
 			enumspb.RESOURCE_EXHAUSTED_CAUSE_CONCURRENT_LIMIT,
```
Comment (yycptt) [inline service/history/workflow/update/registry.go:246]: max total update is not concurrent limit? InvalidArgument or FailedPrecondition?

Comment (alexshtin) [reply service/history/workflow/update/registry.go]: Yes, `FailedPrecondtion` is a better code here.

File: service/history/workflow/update/registry.go [hunk 2]
```diff
@@ -224,14 +234,20 @@ func (r *RegistryImpl) remover(id string) updateOpt {
...
+func (r *RegistryImpl) admit(ctx context.Context) error {
+	if len(r.updates) >= r.maxInFlight() {
 		return serviceerror.NewResourceExhausted(
 			enumspb.RESOURCE_EXHAUSTED_CAUSE_CONCURRENT_LIMIT,
-			fmt.Sprintf("update concurrent in-flight limit has been reached (%v)", max),
+			fmt.Sprintf("update concurrent in-flight limit has been reached (%v)", r.maxInFlight()),
 		)
 	}
+
+	if r.store.GetUpdatesCount(ctx) >= r.maxTotal() {
```
[the remark this answers is not in the document]
Comment (mmcshane) [reply service/history/workflow/update/registry.go:245]: Also, a unit test that hits this, please

File: service/history/workflow/update/registry.go [hunk 3]
```diff
@@ -78,6 +78,7 @@ type (
 		GetAcceptedWorkflowExecutionUpdateIDs(context.Context) []string
 		GetUpdateInfo(ctx context.Context, updateID string) (*persistencespb.UpdateInfo, bool)
 		GetUpdateOutcome(ctx context.Context, updateID string) (*updatepb.Outcome, error)
+		GetUpdatesCount(ctx context.Context) int
```
Comment (mmcshane) [inline service/history/workflow/update/registry.go:81]: These update-specific calls are multiplying. Let's talk separately about if we can keep this interface tighter. Not for this PR though.

**Review by mmcshane — APPROVED** [review]
Approved when unit test for UpdateRegistry is here.

**Review by mmcshane — CHANGES_REQUESTED** [review]

File: service/history/workflow/update/registry_test.go [hunk 4]
```diff
@@ -374,28 +350,82 @@ func TestInFlightLimit(t *testing.T) {
...
+		require.False(t, existed)
+		require.Equal(t, 1, reg.Len())
+
+		_, _, err = reg.FindOrCreate(ctx, "update3")
+		var failedPrecon *serviceerror.FailedPrecondition
+		require.ErrorAs(t, err, &failedPrecon,
+			"update3 should be rejected when limit = 2")
+		require.Equal(t, 1, reg.Len())
+	})
+}
```
Comment (mmcshane) [inline service/history/workflow/update/registry_test.go:415]: I like this test

**Review by mmcshane — APPROVED** [review]
Appreciate your patience with me

**Pull request merged**